src/lib/tls/tls13_pqc/kex_to_kem_adapter.cpp

```diff
@@ -0,0 +1,266 @@
+/**
+ * Adapter that allows using a KEX key as a KEM, using an ephemeral
+ * key in the KEM encapsulation.
+ *
+ * (C) 2023 Jack Lloyd
+ *     2023 Fabian Albert, René Meusel - Rohde & Schwarz Cybersecurity
+ *
+ * Botan is released under the Simplified BSD License (see license.txt)
+ */
+
+#include <botan/internal/kex_to_kem_adapter.h>
+
+#include <botan/internal/fmt.h>
+#include <botan/internal/pk_ops_impl.h>
+#include <botan/internal/stl_util.h>
+
+#include <variant>
+
+#if defined(BOTAN_HAS_DIFFIE_HELLMAN)
+   #include <botan/dh.h>
+   #include <botan/dl_group.h>
+#endif
+
+#if defined(BOTAN_HAS_ECDH)
+   #include <botan/ecdh.h>
+#endif
+
+#if defined(BOTAN_HAS_CURVE_25519)
+   #include <botan/curve25519.h>
+#endif
+
+namespace Botan::TLS {
+
+namespace {
+
+/**
+ * Up-cast a key agreement public key to its concrete algorithm type
+ * to perform low-level operations that are not covered by a generic
+ * interface in Public_Key.
+ */
+std::variant<std::reference_wrapper<const DH_PublicKey>,
+             std::reference_wrapper<const ECDH_PublicKey>,
+             std::reference_wrapper<const Curve25519_PublicKey>>
+as_kex_public_key(const Public_Key& kex_public_key) {
+   BOTAN_ASSERT_NOMSG(kex_public_key.supports_operation(PublicKeyOperation::KeyAgreement));
+
+   try {
+#if defined(BOTAN_HAS_ECDH)
+      if(kex_public_key.algo_name() == "ECDH") {
+         return dynamic_cast<const ECDH_PublicKey&>(kex_public_key);
+      }
```

Comment:
For these it seems better to do dynamic_cast on the pointer; otherwise if someone provides eg `PKCS11_ECDH_PublicKey` which also has an `algo_name` of "ECDH" we'll get a bad dynamic cast, which you do catch, but it's going to be a confusing error message.

Something like

```
if(const auto* ecdh = dynamic_cast<const ECDH_PublicKey*>(&kex_public_key)) {
  return *ecdh;
}
```

Reply:
Fair, I'll fix it.

However, regarding: PKCS #11: That's an issue, no? Shouldn't we be able to use those here?

Reply:
Another issue with `as_kex_public_key()`: It assumes that all algorithms are enabled via `--enable-modules=`. Otherwise the return type `std::variant<>` won't compile. Yikes.

Reply:
In 512e0ffa9117e95a2090294759dc02c1a50b5eaf I got rid of `as_kex_public_key()` entirely, and duplicated the `#if defined(...) \n dynamic_cast<>()` machinery in all downstream helper methods. That way, one can disable (say) Curve25519 at compile time and the KEX-to-KEM adapter still works for the other algorithms.

That's messy. 😞 But its a provisional workaround until #3706 is being tackled.

Reply:
Oof, right. How you did it seems fine given the current limitations of the PK interface.